treematcher.py

```diff
@@ -1,286 +1,178 @@
+            :return: True if node is a duplication, otherwise False.
         """
-        if len(self.find_match(tree, None, maxhits=1))>=1:
-            return True
-        else:
-            return False
+        events = self.cache.get_cached_attr('evoltype', target_node)
+        return(events.count('D'))
 
-    def _parse_constraint(self, node, is_exact=False):
+    def n_speciations(self, target_node):
         """
-        Function for replacing keywords with python code in a query pattern.
-        :param node: The pattern be searched for in the query.
-        :param is_exact: Designates if pattern is in standard Newick format or contains treematcher syntax.
-        :param True is_exact: Pattern is a tree standard Newick format, search for exact match of tree.
-        :param False is_exact:  Pattern contains treematcher syntax and requires keyword to python conversion.
+            Shortcut function to find the number of speciation events at or below a node.
         """
-        node.constraint = node.name
-
-        #builtins = set(PhyloTree.__dict__.keys() + Tree.__dict__.keys())
-        
-
-        if is_exact and node.name != '':
-            node.constraint = "__target.name==" + "'" + str(node.constraint) + "'"
-        else:
-            node.constraint = node.name
-            # turn multiple spaces to single space
-
-            node.constraint = re.sub("\s+", " ", node.constraint)
-
-            if self.cache_flag == True:
-                syntax_tuples = self._syntax_tuples_with_cache
-                #print("cache was flagged")
-            else:
-                syntax_tuples = self._syntax_tuples
-                #print("cache was NOT flagged")
+        events = self.cache.get_cached_attr('evoltype', target_node)
+        return(events.count('S'))
 
-            for keyword, python_code in syntax_tuples:
-                try:
-                    node.constraint = node.constraint.replace(keyword, python_code)
-                except (KeyError, ValueError):
-                    print("Error in syntax dictionary iteration at keyword: " + str(keyword) + "and value: " + python_code)
 
+    # TODO: review next function
 
-            if self._extra_functions is not None:
-                for custom_function in self._extra_functions.keys():
-                    node.constraint = re.sub(custom_function + "\(__target",
-                                             custom_function + "(__target, pattern", node.constraint)
-            if len(self.shortcut_functions)>0:
-                for custom_function in self.shortcut_functions:
-                    node.constraint = re.sub(custom_function + "\(__target",
-                                             custom_function + "(__target, pattern", node.constraint)
-
-            #custom function of stuff tpat=whatever
-
-            if ".lineage" in node.constraint:
-                node.constraint = self.smart_lineage(node.constraint)
-
+    def smart_lineage(self, constraint):
```

Comment:
I still have to review this. I don't understand what you were aiming here @AlishaMechtley. 
In any case, it belongs to PatternSyntax now.